presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java

```diff
@@ -133,115 +200,440 @@ public boolean isFinished()
+                Optional<Block> partialDistinctMask = partialDistinctMasks.get(intermediateIndex);
+                if (partialDistinctMask.isPresent()) {
+                    blocks.add(partialDistinctMask.get());
+                }
+                else if (intermediateIsDistinct.get(intermediateIndex)) {
+                    blocks.add(new RunLengthEncodedBlock(FALSE_SINGLE_VALUE_BLOCK, page.getPositionCount()));
+                }
+                else {
+                    blocks.add(new RunLengthEncodedBlock(inputSingleNullBlocks.get(intermediateIndex), page.getPositionCount()));
+                }
             }
+
+            // add inputs
+            inputs.stream()
+                    .map(AggregationInputChannel::getChannel)
+                    .map(page::getBlock)
+                    .forEach(blocks::add);
+
+            Page outputPage = new Page(page.getPositionCount(), blocks.stream().toArray(Block[]::new));
+
+            // filter output page to selected positions
+            outputPage = allDistinctPositions.selectPositions(outputPage);
+            return Optional.of(outputPage);
         }
 
-        public void evaluate(BlockBuilder blockBuilder)
+        private static SelectedPositions combineDistinctMasks(List<Optional<Block>> partialDistinctMasks)
         {
-            if (step.isOutputPartial()) {
-                aggregation.evaluateIntermediate(blockBuilder);
+            List<Block> presentMasks = partialDistinctMasks.stream()
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .collect(toImmutableList());
+            if (presentMasks.isEmpty()) {
+                return SelectedPositions.empty();
             }
-            else {
-                aggregation.evaluateFinal(blockBuilder);
+
+            int[] distinctPositions = new int[presentMasks.get(0).getPositionCount()];
+            int distinctCount = 0;
+            for (int position = 0; position < distinctPositions.length; position++) {
+                boolean selected = false;
+                for (Block presentMask : presentMasks) {
+                    if (BOOLEAN.getBoolean(presentMask, position)) {
+                        selected = true;
+                        break;
+                    }
+                }
+                if (selected) {
+                    distinctPositions[distinctCount++] = position;
+                }
             }
+            return SelectedPositions.positionsList(distinctPositions, 0, distinctCount);
         }
 
-        public long getEstimatedSize()
+        private Page buildIntermediateOutput(List<Block> intermediateResults)
         {
-            return aggregation.getEstimatedSize();
+            List<Block> blocks = new ArrayList<>(getOutputChannelCount());
+            blocks.add(RowType.INTERMEDIATE.createRowTypeBlock(1));
+            blocks.addAll(intermediateResults);
```

Comment:
Am I understanding correctly that pages with RowType.DISTINCT_PARTIAL have different scheme then pages with RowType.INTERMEDIATE?

RowType.DISTINCT_PARTIAL pages have rowType, followed by one boolean block per distinct aggregation, followed by  one block per input symbol.

For `select count(distinct a), count(distinct b), sum(c) from t`, RowType.DISTINCT_PARTIAL page has 1 + 2 + 3 = 6 blocks: `rowType`, `is_distinct_a:boolean`, `is_distinct_b(boolean)`, `a`, `b`, `c`.

RowType.INTERMEDIATE pages have rowType, followed by one block with partial result per aggregation, followed by some null blocks.

For above query, we have `rowType`, `partial-count(distinct a)`, `partial-count(distinct b)`, `partial-sum(c)`, `null`, `null`, `null`.